boards.txt

```diff
@@ -992,6 +992,12 @@ GenF4.menu.pnum.FEATHER_F405.build.product_line=STM32F405xx
 GenF4.menu.pnum.FEATHER_F405.build.variant=FEATHER_F405
 
 # Upload menu
+GenF4.menu.upload_method.dfuUtil=USB (dfu-util)
+GenF4.menu.upload_method.dfuUtil.upload.tool=dfu_util
+GenF4.menu.upload_method.dfuUtil.upload.protocol=
+GenF4.menu.upload_method.dfuUtil.upload.usbID=0483:df11
+GenF4.menu.upload_method.dfuUtil.upload.address=0x08000000:leave
```

Comment:
I guess this should not be hardcoded if we want flash at another address (ex: user have a bootloader)

Reply:
How would you suggest we make this value dynamic? If the user had a custom bootloader or wanted to flash to a different address, I'd assume they'd be running the command manually from the terminal. (wouldn't it also require the linker script to be updated?) To make it dynamic, are you suggesting a new menu option to choose addresses from?

Reply:
Linker script can already manage this kind of feature as it is preprocessed.
This is already done for BP F103 for example.
flash_offset is currently hardcoded but can be override by a platform.local.txt
Well, this need some investigation :wink:

Reply:
If I understand the example `flash_offset` example, the solution is to rename `upload.address` to `build.flash_offset` to be consistent with the other build flags (although, I feel like I'm missing something).

Reply:
If I understand the `flash_offset` example, the advice would be to change the variable from `upload.address` to `build.address` and likely move the `:leave` flag to the command in platform.txt. I feel like I must be missing something, though.

Originally I was going to rename this to `build. flash_offset` to match the examples, but that would offset where in the program flash it was writing to (which would reduce the overall size of the program).

Reply:
If a custom bootloader (or any other custom data) is present  the max size of the program is reduced.
It assumes that flash_offset  is from the start of the flash.
I wonder if by default the `--dfuse-address {upload.address}` is required maybe use it only if offset is not null.
This is what is done in maple_upload IIRW.

Reply:
Unfortunately, removing `--dfuse-address` doesn't seem to work, with or without the offset being set. dfu-util returns the error:
```
dfu-util: Only DfuSe file version 1.1a is supported
```

From [what I can tell](https://github.com/particle-iot/device-os/issues/706#issuecomment-149600540) this is because of the version of the bootloader is not setting the flash address correctly from the `alt` setting passed to dfu-util. ([error in code](https://sourceforge.net/p/dfu-util/dfu-util/ci/v0.9/tree/src/dfuse.c#l641))

It looks like this might be fixed with dfu-util 1.0, which isn't released yet.

Reply:
> It looks like this might be fixed with dfu-util 1.0, which isn't released yet.

I could not find anything in the changelog you linked about this. Are you perhaps referring to [this commit](https://sourceforge.net/p/dfu-util/dfu-util/ci/64c9f83c60e2fb3f3c6c7c6e6c967b8013b9c6d4/)? If so, I do not think that fixes this, since it only prevents an address of 0 to be treated the same as "no address specified", but when you specify 0 it will actually try to download to address 0, not the start of the area.

I guess dfu-util could be made smarter to default to the start of the first segment (it already parses the DFU altsetting name (e.g. `@Internal Flash  /0x08000000/04*016Kg,01*064Kg,03*128Kg`) to discover the memory layout (needed for STM32-specific erase additions to the DFU protocol). It does not look the current code already does this (assume the first address), though.

Reply:
Originally I was referencing this line:
> Improve alternate interface index matching (Aleks Chakin)

But, after some research, it doesn't look this would help...

I think I figured out why the `alt` flag works for Maple. It seems that maple boards use the 1.1 version of the DFU bootloader protocol ([source](https://github.com/leaflabs/maple-bootloader/blob/master/usb_descriptor.c#L74), [discussion](http://forums.leaflabs.com/forums.leaflabs.com/topic7309.html?id=1410)) and STM32 chips, by default, use 1.1a. The difference is that 1.1 gets the memory addresses from the chip, while the 1.1a protocol requires the host to specify the address explicitly. ([more info](http://dfu-util.sourceforge.net/dfuse.html))

I wish the `dfu-util` would provide an option to override this -- since it already has these addresses from the USB descriptor -- but alas.

The solution preferred by `dfu-util` is to provide a dfu file, instead of a bin, which includes all this information embedded in the file ([source code](https://sourceforge.net/p/dfu-util/dfu-util/ci/master/tree/src/dfu_file.c#l246)). STM32 provides a [bin -> dfu file converter](https://docs.ghielectronics.com/hardware/loaders/stm32-bootloader.html), however, that program also requires the starting flash address to work.

**Conclusion**
So, ultimately, it seems we're left with the requirement to send the flash address with the command (unless I'm missing something). It would be _possible_ to write a script which uses `dfu-util list` to parse the flash addresses first, but that sounds like overkill and error-prone.

Reply:
> I think I figured out why the alt flag works for Maple. It seems that maple boards use the 1.1 version of the DFU bootloader protocol (source , discussion) and STM32 chips, by default, use 1.1a. The difference is that 1.1 gets the memory addresses from the chip, while the 1.1a protocol requires the host to specify the address explicitly. (more info)

I actually think it is a little bit different. In the official 1.1 protocol, there is no concept of addresses. You just pass a blob of binary to the bootloader, and the bootloader will decide where to place it. The only way to choose between different placements is using different alts (but again, there is no standard way to map these onto addresses).

The "1.1a" version of the protocol (which, AFAICS is not an official standard but something STM came up with and labeled with a non-standard version number which can now no longer be used for future standard versions) does things differently: You must explicitly tell the bootloader where the binary should be uploaded (in addition, sectors must be explicitly erased before writing, which the DFU standard states should be handled by the bootloader automatically). For this, the programmer needs info about addresses and sector layout, which is encoded in the "name" of the alt (e.g. `@Internal Flash /0x08000000/04*016Kg,01*064Kg,03*128Kg` encodes the name, address and sector layout in an STM-specific format).

> I wish the dfu-util would provide an option to override this -- since it already has these addresses from the USB descriptor -- but alas.

Yes, I think all info should be present and such a change to dfu-util is feasible (probably not even very complicated). I would want to have a stab at this, but it's enough of a priority for me or my customer to actually free up time for this, I'm afraid.

> So, ultimately, it seems we're left with the requirement to send the flash address with the command (unless I'm missing something). It would be possible to write a script which uses dfu-util list to parse the flash addresses first, but that sounds like overkill and error-prone.

Agreed.

Reply:
If you want to have stab at implementing the auto address selection in dfu-util, let me know and I'll point you to some things I found while reading the source yesterday :-)

Reply:
@matthijskooijman this could be interesting to know 😃